upgrade macos version for support c++11 (#156)
joerick/cibuildwheel: README.md

- The project directory is mounted in the running Docker instance as `/project`, the output directory for the wheels as `/output`. In general, this is handled transparently by `cibuildwheel`. For a more finegrained level of control however, the root of the host file system is mounted as `/host`, allowing for example to access shared files, caches, etc. on the host file system.  Note that this is not available on CircleCI due to their Docker policies.
- Alternative dockers images can be specified with the `CIBW_MANYLINUX1_X86_64_IMAGE` and `CIBW_MANYLINUX1_I686_IMAGE` options to allow for a custom, preconfigured build environment for the Linux builds. See [below](#options) for more details.

### Modern C++ standards

Building Python wheels with modern C++ (C++11+) requires a few tricks.

Python 2.7 wheels cannot use C++17 at all, because Python 2.7 uses the `register` keyword, which is prohibited in C++17 standard.

#### Linux

> For `manylinux1` images, only C++11 standards is supported. C++14 and C++17 will be supported with manylinux2010 image (coming soon!).

Author: change manylinux2010 to `manylinux2010`

#### MacOS

To get C++11 and C++14 support, set `MACOSX_DEPLOYMENT_TARGET` variable to `10.9`.

To get C++17 support need to set `MACOSX_DEPLOYMENT_TARGET` variable to `10.13` or `10.14`. (`10.13` supports C++17 partially, e.g. filesystem header is in experimental: `#include <filesystem>` -> `#include <experimental/filesystem>`)

#### Windows

> Visual C++ for Python 2.7 do not support modern standards of C++.

Author: this configuration do not support. Here is example how get this possibility, but I do not understand it yet: https://github.com/pybind/python_example

Reviewer: ```suggestion
Visual C++ for Python 2.7 does not support modern standards of C++.
```

> When building on Appveyor, you'll need to use the 'Visual Studio 2019' image. Note that Python 2.7 isn't supported on this image - you should skip it using `CIBW_SKIP=cp27-win*`.

Author: For c++14 and c++17 both images should work: 'Visual Studio 2017' or 'Visual Studio 2019'

Reviewer 2: ```suggestion
When building on Appveyor, you'll need to use either the 'Visual Studio 2017' or 'Visual Studio 2019' image. Note that Python 2.7 isn't supported on these images - you should skip it using `CIBW_SKIP=cp27-win*`.
```

- The project directory is mounted in the running Docker instance as `/project`, the output directory for the wheels as `/output`. In general, this is handled transparently by `cibuildwheel`. For a more finegrained level of control however, the root of the host file system is mounted as `/host`, allowing for example to access shared files, caches, etc. on the host file system.  Note that this is not available on CircleCI due to their Docker policies.
- Alternative dockers images can be specified with the `CIBW_MANYLINUX1_X86_64_IMAGE` and `CIBW_MANYLINUX1_I686_IMAGE` options to allow for a custom, preconfigured build environment for the Linux builds. See [below](#options) for more details.

Moden C++ standards
-------------------
Python 2.7 use `register` keyword which is prohibited in c++17 standard. 

Linux
=====
For `manulinux1` images only c++11 standards is supported. c++14 and c++17 will be supported with manylinux2010 image

MacOS
=====

> To get c++11 and c++14 upport need to set `MACOSX_DEPLOYMENT_TARGET` variable to `10.9` value

Reviewer 3: ```suggestion
To get c++11 and c++14 support need to set `MACOSX_DEPLOYMENT_TARGET` variable to `10.9` value
```

Do newer macos versions work as well? Or just `10.9`?

Author: 10.9 is minimal version of macos on which package will work.

- The project directory is mounted in the running Docker instance as `/project`, the output directory for the wheels as `/output`. In general, this is handled transparently by `cibuildwheel`. For a more finegrained level of control however, the root of the host file system is mounted as `/host`, allowing for example to access shared files, caches, etc. on the host file system.  Note that this is not available on CircleCI due to their Docker policies.
- Alternative dockers images can be specified with the `CIBW_MANYLINUX_X86_64_IMAGE` and `CIBW_MANYLINUX_I686_IMAGE` options to allow for a custom, preconfigured build environment for the Linux builds. See [below](#options) for more details.

> ### Modern C++ standards

Reviewer 4: I'm a little bit in doubt on where we should add this, and in which amount of detail. I agree modern C++ versions are important, but `cibuildwheel`is independent of the language/compilation that generates/compiles the C extensions.

So in that case, it might be better to create a page with FAQ on different programming languages. And rather than explaining everything ourselves, we should link to the relevant resources so that users can find out more.

Building Python wheels with modern C++ (C++11) requires a few tricks.

> Python 2.7 wheels cannot use C++17 at all, because Python 2.7 uses the `register` keyword, which is prohibited in C++17 standard.

Author: I think that all technical details should be moved to documentation and only linked from readme.

> Python 2.7 wheels cannot use C++17 at all, because Python 2.7 uses the `register` keyword, which is prohibited in C++17 standard.

Reviewer 4: Yes, definitely. But this comment is about the content, not about how to merge this into the new docs.

> Python 2.7 wheels cannot use C++17 at all, because Python 2.7 uses the `register` keyword, which is prohibited in C++17 standard.

Reviewer 5: `-Dregister=` or equivalent (that is, defining `register` as early as possible as a macro which expands to nothing) should also be an effective workaround for this unfortunate change in C++17.

> Python 2.7 wheels cannot use C++17 at all, because Python 2.7 uses the `register` keyword, which is prohibited in C++17 standard.

Reviewer 4: Good point! I think I was planning to do that, but then the compiler had a flag to just do so, so that was easier :-)

Bottom line: a lot of things are possible, but are not really supported in a straightforward way and need some hacks here and there. But I think we should be careful formulating what's not possible.

> Python 2.7 wheels cannot use C++17 at all, because Python 2.7 uses the `register` keyword, which is prohibited in C++17 standard.

Author: I change code to test also c++17 against python 2.7. But not on windows.

#### Linux

> `manylinux2010` supports all C++ standards. If you would like to build `manylinux1` images, then your project is limited to C++11 and earlier. 

Author: I see, but I write this text before manylinux2010 is merged. In few days I will rebase it on current master and move most of changes to docs.

> `manylinux2010` supports all C++ standards. If you would like to build `manylinux1` images, then your project is limited to C++11 and earlier. 

Reviewer 4: This has got nothing to do with `manylinux2010`. This is about how you can make C++11/C++14 wheels in `manylinux2014`, through some tricks.

> `manylinux2010` supports all C++ standards. If you would like to build `manylinux1` images, then your project is limited to C++11 and earlier. 

Author: I do not understand. I can build c++11 on `manylinux1`. `manylinux2014` is still not supported by `cibuildwheel` and why its need special tricks to use these standards.

> `manylinux2010` supports all C++ standards. If you would like to build `manylinux1` images, then your project is limited to C++11 and earlier. 

Reviewer 4: Sorry. that was a typo. I meant to say: "This is about how you can make C++11/C++14 wheels in `manylinux1`, through some tricks."

- The project directory is mounted in the running Docker instance as `/project`, the output directory for the wheels as `/output`. In general, this is handled transparently by `cibuildwheel`. For a more finegrained level of control however, the root of the host file system is mounted as `/host`, allowing for example to access shared files, caches, etc. on the host file system.  Note that this is not available on CircleCI due to their Docker policies.
- Alternative dockers images can be specified with the `CIBW_MANYLINUX_X86_64_IMAGE` and `CIBW_MANYLINUX_I686_IMAGE` options to allow for a custom, preconfigured build environment for the Linux builds. See [below](#options) for more details.

### Modern C++ standards

Building Python wheels with modern C++ (C++11) requires a few tricks.

Python 2.7 wheels cannot use C++17 at all, because Python 2.7 uses the `register` keyword, which is prohibited in C++17 standard.

#### Linux

> `manylinux2010` support all c++ standards. If you would like to build `manylinux1` images then your project is limited to c++11 and earlier. 

Reviewer: ```suggestion
`manylinux2010` supports all C++ standards. If you would like to build `manylinux1` images, then your project is limited to C++11 and earlier. 
```

#### MacOS

To get C++11 and C++14 support, set `MACOSX_DEPLOYMENT_TARGET` variable to `"10.9"`.

> To get C++17 support need to set `MACOSX_DEPLOYMENT_TARGET` variable to `"10.13"` or `"10.14"`. (`"10.13"` supports C++17 partially, e.g. filesystem header is in experimental: `#include <filesystem>` -> `#include <experimental/filesystem>`)

Reviewer: ```suggestion
To get C++17 support, set `MACOSX_DEPLOYMENT_TARGET` variable to `"10.13"` or `"10.14"`. (`"10.13"` supports C++17 partially, e.g. the filesystem header is in experimental: `#include <filesystem>` -> `#include <experimental/filesystem>`)
```

Python tags look like `cp27` `cp34` `cp35` `cp36` `cp37`

> 

Reviewer 2: Could we do one of those identifier migrations like how @YannickJadoul did the `manylinux1`->`manylinux` warnings?

Author: added

Author: added
